### PR TITLE
fixing transitive order of privates

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Bugfix: Fix skipping binaries because of transitive ``private`` requirements
Docs: omit

@tags: slow

Close: #4970

### CODE REVIEW

**Comment by memsharded** [comment]
Hi @theodelrieu @DoDoENT @puetzk @thorntonryan @jallenwork

Please find here a new patch for Conan 1.14.4 for the processing of ``private``,  ``build-requires`` and mixed types dependencies.
It would be great if you could try it and give it a go with your projects, to check before merging and releasing.

It happens that the build-requires and private usage is way more complicated that what the model was designed for. We are trying to propose fixes and improvements to achieve both a safe dependency resolution (linking with the right dependency, conflicting if not defined) and those complex graphs Conan didn't intended to support initially (but we are doing our best to try to adapt).

Many thanks!!

**Comment by theodelrieu** [comment]
All I can say is that nothing broke on our end :sweat_smile:

**Comment by jallenwork** [comment]
This fixes our larger reproducer. Seems to be all good on our end. Thanks for putting in the time and effort to expand the requires/build_requires functionality for the more complex cases! This is really helpful. 😄

**Comment by DoDoENT** [comment]
I've tried that patch on both may old codebase and new one (after making update as suggested in your [comment](https://github.com/conan-io/conan/issues/4753#issuecomment-481663245). The old one reports conflict, just like with conan v1.13+ and as demonstrated in minimum sample attached to the same issue and the new one works the same as with conan v1.14.3. Basically, I haven't seen any changes to that part of my package management from v1.14.3 to v1.14.4.

Basically, for us its the same as for @theodelrieu - nothing broke 😅

**Comment by thorntonryan** [comment]
@memsharded ,

Oye, I apologize for being late to the party. I had an example thrown together last night, but it was getting late so I opted to committing it in the morning... looks like that was a mistake.

As for my original example with our shared COM dlls,  I think the solution is to use `requirements` instead of `build_requirements`, which will get the override behavior we're after.  
_Original example discussed here https://github.com/conan-io/conan/issues/4753#issuecomment-477774018_

So in that specific case, I think we were misusing `build_requires` and there's not a strong enough case to argue for the use of "private" `requires` with these COM dlls.

However, in the same project, we're also using the header only library Boost, and I believe _that_ raises some questions about the way "private" requirements are resolved that are worth considering.

So, yesterday, I worked with @puetzk to come up with a more illustrative example of some potential issues we see with the new approach to "private" requirements in 1.14.4.

The primary concern is that the 1.14.4 "private" requirements could result in One Definition Rule (ODR) violations if a shared package uses header only or static libraries in such a way that they are `fully embedded and hidden from consumers of the package.` and `not exposed through the public interface` _and_ the consuming app also uses these same header only or static libraries.

My new examples attempt to illustrate this by using "mock" packages of two popular libraries, Boost and zlib.

When using Boost in header only mode, all users of Boost must maintain `full_version_mode` compatibility, otherwise they risk violating ODR.

On the other hand, libraries like zlib  make very strong semver ABI guarantees. A consuming app using a newer semver compatible version of zlib could, in theory, override the version of zlib used by one of its required packages.

But in both cases, Conan 1.14.4 resolves these requirements and "private" requirements in such a way that both versions of Boost and zlib are present in the consuming package... which seems problematic.

You can find that here:
https://github.com/thorntonryan/conan-requirements-overriding-change-in-behavior
_Specifically added here, thorntonryan/conan-requirements-overriding-change-in-behavior@6d000202b4cdd44c60664ca5e0f44540ea9767c8_

We can work around this issue by promoting these "private" requirements to full on requirements, and then the overriding will work, but it seems a little odd to make "public" requirements on Boost / zlib when they're not at all exposed through the public API of our shared library.

**Comment by theodelrieu** [comment]
Hi, I should have tried on my side project beforehand... I get the following assert:

>ERROR: printing with scope==virtual

With 1.14.4, sorry about that!

You can reproduce by checking out the `conan_bug` branch [here](https://github.com/theodelrieu/playground/tree/conan_bug) and running `conan workspace install conanws.yml`.

**Comment by memsharded** [comment]
Hi @theodelrieu 

https://github.com/conan-io/conan/pull/5056 has a possible fix to this issue, to be included if ok in 1.14.5 release. 

If you could please run that branch from source against your project, to make sure everything is ok, that would be great. Thanks!

**Comment by theodelrieu** [comment]
Seems to work, thanks! :)
